tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -3422,6 +3437,259 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
+      theBuilder.createStore(expId, exponentFloat);
+      return theBuilder.createCompositeExtract(argTypeId, frexp, {0});
+    }
+  }
+
+  // For matrix argument types.
+  {
+    uint32_t rowCount = 0, colCount = 0;
+    if (TypeTranslator::isMxNMatrix(argType, nullptr, &rowCount, &colCount)) {
+      const auto floatId = theBuilder.getFloat32Type();
+      const auto expTypeId = theBuilder.getVecType(intId, colCount);
+      const auto colTypeId = theBuilder.getVecType(floatId, colCount);
+      const auto frexpStructTypeId = theBuilder.getStructType(
+          {colTypeId, expTypeId}, "FrexpStructType", {"mantissa", "exponent"});
+      llvm::SmallVector<uint32_t, 4> exponents;
+      llvm::SmallVector<uint32_t, 4> mantissas;
+      for (uint32_t i = 0; i < rowCount; ++i) {
+        const auto curRow =
+            theBuilder.createCompositeExtract(colTypeId, argId, {i});
+        const auto frexp = theBuilder.createExtInst(
+            frexpStructTypeId, glslInstSetId, GLSLstd450::GLSLstd450FrexpStruct,
+            {curRow});
+        const auto exponentInt =
+            theBuilder.createCompositeExtract(expTypeId, frexp, {1});
+
+        // Since the SPIR-V instruction returns an int, and the intrinsic HLSL
+        // expects a float, an conversion must take place before writing the
+        // results.
+        const auto exponentFloat = theBuilder.createUnaryOp(
+            spv::Op::OpConvertSToF, colTypeId, exponentInt);
+        exponents.push_back(exponentFloat);
+        mantissas.push_back(
+            theBuilder.createCompositeExtract(colTypeId, frexp, {0}));
+      }
+      const auto exponentsResultId =
+          theBuilder.createCompositeConstruct(returnTypeId, exponents);
+      theBuilder.createStore(expId, exponentsResultId);
+      return theBuilder.createCompositeConstruct(returnTypeId, mantissas);
+    }
+  }
+
+  emitError("Unknown argument type passed to Frexp function.");
+  return 0;
+}
+
+uint32_t SPIRVEmitter::processIntrinsicClip(const CallExpr *callExpr) {
+  // Discards the current pixel if the specified value is less than zero.
+  // TODO: If the argument can be const folded and evaluated, we could
+  // potentially avoid creating a branch. This would be a bit challenging for
+  // matrix/vector arguments.
+
+  assert(callExpr->getNumArgs() == 1u);
+  const Expr *arg = callExpr->getArg(0);
+  const auto argType = arg->getType();
+  const auto boolType = theBuilder.getBoolType();
+  uint32_t condition = 0;
+
+  // Could not determine the argument as a constant. We need to branch based on
+  // the argument. If the argument is a vector/matrix, clipping is done if *any*
+  // element of the vector/matrix is less than zero.
+  const uint32_t argId = doExpr(arg);
+
+  QualType elemType = {};
+  uint32_t elemCount = 0, rowCount = 0, colCount = 0;
+  if (TypeTranslator::isScalarType(argType)) {
+    const auto zero = getValueZero(argType);
+    const auto cmpType = boolType;
```

Comment:
Why not use `boolType` directly in the following? It seems there is no need to create a new  short lived `cmpType`.

Reply:
Done